include/stl2/view/drop.hpp

```diff
@@ -0,0 +1,228 @@
+// cmcstl2 - A concept-enabled C++ standard library
+//
+//  Copyright Casey Carter
+//  Copyright Eric Niebler
+//  Copyright Christopher Di Bella
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/caseycarter/cmcstl2
+//
+#ifndef STL2_VIEW_DROP_HPP
+#define STL2_VIEW_DROP_HPP
+
+#include <stl2/detail/fwd.hpp>
+#include <stl2/detail/meta.hpp>
+#include <stl2/detail/semiregular_box.hpp>
+#include <stl2/detail/iterator/concepts.hpp>
+#include <stl2/detail/functional/invoke.hpp>
+#include <stl2/detail/range/access.hpp>
+#include <stl2/detail/range/concepts.hpp>
+#include <stl2/detail/view/view_closure.hpp>
+#include <stl2/view/all.hpp>
+#include <stl2/view/view_interface.hpp>
+
+STL2_OPEN_NAMESPACE {
+	namespace ext {
+		template <Range R>
+		requires
+			View<R>
+		struct drop_view;
+
+		template <Range R>
+		requires
+			View<R>
+		class __drop_view : public view_interface<drop_view<R>> {
+			using D = iter_difference_t<iterator_t<R>>;
+
+			template <bool Const>
+			struct __sentinel;
+		public:
+			__drop_view() = default;
+
+			constexpr __drop_view(R base, D count)
+				: base_(std::move(base)),
+				  count_(count)
+			{}
+
+			template <Range O>
+			requires
+				ext::ViewableRange<O> &&
+				_ConstructibleFromRange<R, O>
+			constexpr __drop_view(O&& o, D count)
+				: base_(view::all(std::forward<O>(o))),
+				  count_(count)
+			{}
+
+			constexpr R base() const
+			{
+				return base_;
+			}
+
+			constexpr auto begin()
+			{
+				return begin_impl(*this);
+			}
+
+			constexpr auto begin() const
+			requires
+				Range<R const>
+			{
+				return begin_impl(*this);
+			}
+
+			constexpr auto begin()
+			requires
+				SizedRange<R>
+			{
+				return begin_impl(*this);
+			}
+
+			constexpr auto begin() const
+			requires
+				SizedRange<R const>
+			{
+				return begin_impl(*this);
+			}
+
+			constexpr auto end()
+			{
+				return __sentinel<SimpleView<R>>(__stl2::end(base_));
+			}
+
+			constexpr auto end() const
+			requires
+				Range<R const>
+			{
+				return __sentinel<true>(__stl2::end(base_));
+			}
+
+			constexpr auto size() const
+			requires
+				SizedRange<R>
+			{
+				return __stl2::distance(base_) < count_ ? 0 : __stl2::distance(base_) - count_;
+			}
+		private:
+			R base_;
+			D count_;
+
+			template <class X>
+			static constexpr decltype(auto) begin_impl(X&& x)
+			{
+				return __stl2::next(__stl2::begin(x.base_), x.count_);
+			}
+
+			template <class X>
+			static constexpr decltype(auto) begin_impl(X&& x)
+			requires
+				SizedRange<R>
+			{
+				D const dist = __stl2::distance(x.base_);
+				return (__stl2::next(__stl2::begin(x.base_), dist < x.count_ ? dist : x.count_));
```

Comment:
An `STL2_EXPECT(count_ >= 0);` before the `next` call would likely improve codegen by telling the optimizer that it can eliminate the code that handles negative counts for bidirectional iterators in `next`.

Reply:
And now the nasty part: this is `O(1)` for random access ranges, but nothing weaker. For weaker underlying ranges, we need to cache the first iterator in a `detail::non_propagating_cache`. (`reverse_view` is a good example.) Consequently, the `const` overloads of `begin` and `end` need to be enabled only when `R` models `RandomAccessRange`.

Reply:
I think this and the `!SimpleView` requirement on the non-`const` overload are altogether a bit too much. `std::list` isn't a `RandomAccessRange` and it also isn't a `SimpleView`, but it should be usable with `drop`. I'm partial to making a mutable `non_propagating_cache`, rather than inheriting from it.

Reply:
The point of the `!SimpleView` constraint on the non-`const` overloads is to avoid instantiating them when they would be equivalent to the `const` overloads. Adding the `RandomAccessRange` constraint to the `const` overloads needs to be reflected in the non-`const` constraint to maintain that equivalence: they should be constrained with `!(SimpleView<R> && RandomAccessRange<R>)`.

Inheriting from `non_propagating_cache` makes it EBO away when it's disabled and therefore empty, which is nice. `mutable` is a recipe for violating [[res.on.data.races]](http://eel.is/c++draft/res.on.data.races) - updating a `mutable` cache from a `const` member function means that simultaneous calls to e.g. `begin` can cause data races.